File: Sources/structly/04_linked_list/LongestStreak.swift
//
//  LongestStreak.swift
//  MyCLI
//
//  Created by new on 3/22/26.
//

func longestStreak<T: Equatable>(_ head: Node<T>?) -> Int {
    func longest(_ node: Node<T>?) -> (Int, Int) {
        guard let node else {
            return (0, 0)
        }

        var (look_ahead_length, global_max) = longest(node.next)
        let res = 1 + (node.val == node.next?.val ? look_ahead_length : 0)

        global_max = max(global_max, res)

        return (res, global_max)
    }

    let (_, global_max) = longest(head)

    return global_max;
}

File: Sources/structyC/04_linked_list/longest_streak.cpp
//
//  longest_streak.cpp
//  MyCLI
//
//  Created by new on 3/22/26.
//
#include <algorithm>

class Node {
public:
    int val;
    Node* next;

    Node(int initialVal) {
        val = initialVal;
        next = nullptr;
    }
};

int longestStreak(Node* head) {
    int max_streak = 0;
    int prev_value = 42;
    int current_streak = 0;
    Node *node = head;
    while (node != nullptr) {
        if (node == head || node->val != prev_value) {
            current_streak = 1;
        } else {
            current_streak += 1;
        }

        prev_value = node->val;
        max_streak = std::max(max_streak, current_streak);

        node = node->next;
    }
    return max_streak;
}
